File: plugins/methods/src/com/intellij/methods/vfs/MethodsVirtualFile.java
package com.intellij.methods.vfs;

import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;

/**
 * Lightweight virtual file storing generated content in-memory.
 */
public final class MethodsVirtualFile extends VirtualFile {
  private final String myPath;
  private final String myName;
  private volatile byte[] myContent = new byte[0];
  private volatile long myStamp = 0L;

  MethodsVirtualFile(@NotNull String path) {
    myPath = path;
    int idx = path.lastIndexOf('/');
    myName = idx >= 0 ? path.substring(idx + 1) : path;
  }

  void updateContent(byte[] data, long stamp) {
    myContent = data;
    myStamp = stamp;
  }

  private String fullPath() {
    return MethodsVirtualFileSystem.PROTOCOL + "://" + myPath;
  }

  @Override
  public @NotNull String getName() {
    return myName;
  }

  @Override
  public @NotNull VirtualFileSystem getFileSystem() {
    return MethodsVirtualFileSystem.getInstance();
  }

  @Override
  public @NotNull String getPath() {
    return fullPath();
  }

  @Override
  public boolean isWritable() {
    return false;
  }

  @Override
  public boolean isDirectory() {
    return false;
  }

  @Override
  public boolean isValid() {
    return true;
  }

  @Override
  public VirtualFile getParent() {
    return null;
  }

  @Override
  public VirtualFile[] getChildren() {
    return EMPTY_ARRAY;
  }

  @Override
  public @NotNull InputStream getInputStream() {
    byte[] data = contentsToByteArray();
    return new ByteArrayInputStream(data);
  }

  @Override
  public byte @NotNull [] contentsToByteArray() {
    byte[] data = myContent;
    if (data.length == 0) {
      data = MethodsVirtualFileSystem.buildContent(myPath);
      updateContent(data, System.currentTimeMillis());
    }
    return data;
  }

  @Override
  public long getTimeStamp() {
    return myStamp;
  }

  @Override
  public long getLength() {
    return contentsToByteArray().length;
  }

  @Override
  public void refresh(boolean asynchronous, boolean recursive, @Nullable Runnable postRunnable) {
  }

  @Override
  public @NotNull OutputStream getOutputStream(Object requestor, long newModificationStamp, long newTimeStamp) throws IOException {
    throw new IOException("Read only");
  }

  @Override
  public long getModificationStamp() {
    return myStamp;
  }

  @Override
  public @NotNull byte[] getOutputStreamContent() {
    return myContent;
  }
}

File: plugins/methods/src/com/intellij/methods/vfs/MethodsVirtualFile.kt
package com.intellij.methods.vfs

import com.intellij.openapi.vfs.VirtualFile
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

/**
 * Lightweight virtual file storing generated content in-memory.
 * Designed to be extracted into a plugin in the future.
 */
class MethodsVirtualFile(private val myPath: String) : VirtualFile() {
  private val myName: String = myPath.substringAfterLast('/')
  @Volatile private var myContent: ByteArray = ByteArray(0)
  @Volatile private var myStamp: Long = 0

  fun updateContent(data: ByteArray, stamp: Long) {
    myContent = data
    myStamp = stamp
  }

  private fun fullPath(): String = "${MethodsVirtualFileSystem.PROTOCOL}://$myPath"

  override fun getName(): String = myName
  override fun getFileSystem() = MethodsVirtualFileSystem.getInstance()
  override fun getPath(): String = fullPath()
  override fun isWritable(): Boolean = false
  override fun isDirectory(): Boolean = false
  override fun isValid(): Boolean = true
  override fun getParent(): VirtualFile? = null
  override fun getChildren(): Array<VirtualFile> = EMPTY_ARRAY
  override fun getInputStream(): InputStream = ByteArrayInputStream(contentsToByteArray())

  override fun contentsToByteArray(): ByteArray {
    var data = myContent
    if (data.isEmpty()) {
      data = MethodsVirtualFileSystem.buildContent(myPath)
      updateContent(data, System.currentTimeMillis())
    }
    return data
  }

  override fun getTimeStamp(): Long = myStamp
  override fun getLength(): Long = contentsToByteArray().size.toLong()
  override fun refresh(asynchronous: Boolean, recursive: Boolean, postRunnable: Runnable?) {}
  override fun getOutputStream(requestor: Any?, newModificationStamp: Long, newTimeStamp: Long): OutputStream {
    throw IOException("Read only")
  }
  override fun getModificationStamp(): Long = myStamp
}
